diff --git a/server.go b/server.go
index 00d7560..98f0424 100644
--- a/server.go
+++ b/server.go
@@ -339,7 +339,7 @@ func (s *Server) sendWebPush(ctx context.Context, sub *webpush.Subscription, vap
 		},
 		VAPIDPublicKey:  s.webPush.VAPIDKeys.Public,
 		VAPIDPrivateKey: s.webPush.VAPIDKeys.Private,
-		Subscriber:      "https://soju.im",
+		Subscriber:      "hello@pico.sh",
 		TTL:             7 * 24 * 60 * 60, // seconds
 		Urgency:         urgency,
 		RecordSize:      2048,
