diff --git a/downstream.go b/downstream.go
index b627450..b845ecd 100644
--- a/downstream.go
+++ b/downstream.go
@@ -3341,6 +3341,9 @@ func (dc *downstreamConn) handleMessageRegistered(ctx context.Context, msg *irc.
 				network.pushTargets.Del(target)
 			}
 			go network.broadcastWebPush(&irc.Message{
+				Tags: irc.Tags{
+					"bouncerNetwork": strconv.FormatInt(dc.network.ID, 10),
+				},
 				Command: "MARKREAD",
 				Params:  []string{target, timestampStr},
 			})
@@ -3694,6 +3697,10 @@ func (dc *downstreamConn) handleMessageRegistered(ctx context.Context, msg *irc.
 				}}
 			}
 
+			if filterCommands, filterCommandExists := rawKeys["filterCommands"]; filterCommandExists {
+				endpoint = fmt.Sprintf("%s;%s", endpoint, filterCommands)
+			}
+
 			newSub := database.WebPushSubscription{
 				Endpoint: endpoint,
 			}
diff --git a/server.go b/server.go
index f1f56d7..81c8bd6 100644
--- a/server.go
+++ b/server.go
@@ -10,6 +10,7 @@ import (
 	"net/http"
 	"net/netip"
 	"runtime/debug"
+	"strings"
 	"sync"
 	"sync/atomic"
 	"syscall"
@@ -340,6 +341,15 @@ func (s *Server) sendWebPush(ctx context.Context, sub *webpush.Subscription, vap
 	ctx, cancel := context.WithTimeout(ctx, 15*time.Second)
 	defer cancel()
 
+	if strings.Contains(sub.Endpoint, ";") {
+		parts := strings.Split(sub.Endpoint, ";")
+		sub.Endpoint = parts[0]
+
+		if !strings.Contains(parts[1], msg.Command) {
+			return nil
+		}
+	}
+
 	var urgency webpush.Urgency
 	switch msg.Command {
 	case "PRIVMSG", "NOTICE", "INVITE":
@@ -352,7 +362,7 @@ func (s *Server) sendWebPush(ctx context.Context, sub *webpush.Subscription, vap
 		HTTPClient:      webPushHTTPClient,
 		VAPIDPublicKey:  s.webPush.VAPIDKeys.Public,
 		VAPIDPrivateKey: s.webPush.VAPIDKeys.Private,
-		Subscriber:      "https://soju.im",
+		Subscriber:      "hello@pico.sh",
 		TTL:             7 * 24 * 60 * 60, // seconds
 		Urgency:         urgency,
 		RecordSize:      2048,
diff --git a/upstream.go b/upstream.go
index 37cf508..8f6ff03 100644
--- a/upstream.go
+++ b/upstream.go
@@ -771,6 +771,7 @@ func (uc *upstreamConn) handleMessage(ctx context.Context, msg *irc.Message) err
 		}
 
 		if sendWebPush {
+			msg.Tags["bouncerNetwork"] = strconv.FormatInt(uc.network.ID, 10)
 			timestamp, err := time.Parse(xirc.ServerTimeLayout, msg.Tags["time"])
 			if err != nil {
 				// Cannot fail, was set above if unset or invalid.
@@ -1760,6 +1761,7 @@ func (uc *upstreamConn) handleMessage(ctx context.Context, msg *irc.Message) err
 		})
 
 		if weAreInvited {
+			msg.Tags["bouncerNetwork"] = strconv.FormatInt(uc.network.ID, 10)
 			go uc.network.broadcastWebPush(msg)
 		}
 	case irc.RPL_INVITING:
