diff --git a/downstream.go b/downstream.go
index f383212..35082d4 100644
--- a/downstream.go
+++ b/downstream.go
@@ -2947,6 +2947,9 @@ func (dc *downstreamConn) handleMessageRegistered(ctx context.Context, msg *irc.
 				network.pushTargets.Del(target)
 			}
 			go network.broadcastWebPush(&irc.Message{
+				Tags: irc.Tags{
+					"bouncerNetwork": strconv.FormatInt(dc.network.ID, 10),
+				},
 				Command: "MARKREAD",
 				Params:  []string{target, timestampStr},
 			})
@@ -3198,6 +3201,10 @@ func (dc *downstreamConn) handleMessageRegistered(ctx context.Context, msg *irc.
 				}}
 			}
 
+			if filterCommands, filterCommandExists := rawKeys["filterCommands"]; filterCommandExists {
+				endpoint = fmt.Sprintf("%s;%s", endpoint, filterCommands)
+			}
+
 			newSub := database.WebPushSubscription{
 				Endpoint: endpoint,
 			}
diff --git a/server.go b/server.go
index 552104e..4b2ffe6 100644
--- a/server.go
+++ b/server.go
@@ -10,6 +10,7 @@ import (
 	"net"
 	"net/http"
 	"runtime/debug"
+	"strings"
 	"sync"
 	"sync/atomic"
 	"time"
@@ -325,6 +326,15 @@ func (s *Server) sendWebPush(ctx context.Context, sub *webpush.Subscription, vap
 	ctx, cancel := context.WithTimeout(ctx, 15*time.Second)
 	defer cancel()
 
+	if strings.Contains(sub.Endpoint, ";") {
+		parts := strings.Split(sub.Endpoint, ";")
+		sub.Endpoint = parts[0]
+
+		if !strings.Contains(parts[1], msg.Command) {
+			return nil
+		}
+	}
+
 	var urgency webpush.Urgency
 	switch msg.Command {
 	case "PRIVMSG", "NOTICE", "INVITE":
@@ -339,7 +349,7 @@ func (s *Server) sendWebPush(ctx context.Context, sub *webpush.Subscription, vap
 		},
 		VAPIDPublicKey:  s.webPush.VAPIDKeys.Public,
 		VAPIDPrivateKey: s.webPush.VAPIDKeys.Private,
-		Subscriber:      "https://soju.im",
+		Subscriber:      "hello@pico.sh",
 		TTL:             7 * 24 * 60 * 60, // seconds
 		Urgency:         urgency,
 		RecordSize:      2048,
diff --git a/upstream.go b/upstream.go
index ca56a82..8ff0a55 100644
--- a/upstream.go
+++ b/upstream.go
@@ -682,6 +682,7 @@ func (uc *upstreamConn) handleMessage(ctx context.Context, msg *irc.Message) err
 		}
 
 		if highlight || directMessage {
+			msg.Tags["bouncerNetwork"] = strconv.FormatInt(uc.network.ID, 10)
 			go uc.network.broadcastWebPush(msg)
 			if timestamp, err := time.Parse(xirc.ServerTimeLayout, string(msg.Tags["time"])); err == nil {
 				uc.network.pushTargets.Set(bufferName, timestamp)
@@ -1611,6 +1612,7 @@ func (uc *upstreamConn) handleMessage(ctx context.Context, msg *irc.Message) err
 		})
 
 		if weAreInvited {
+			msg.Tags["bouncerNetwork"] = strconv.FormatInt(uc.network.ID, 10)
 			go uc.network.broadcastWebPush(msg)
 		}
 	case irc.RPL_INVITING:
